src/shim.c

```diff
@@ -151,100 +155,224 @@ err_exit(const char *format, ...)
 	exit(EXIT_FAILURE);
 }
 
-/*!
- * Construct message in the proxy ctl rpc protocol format
- * Proxy control message format:
- *
- * | length  | Reserved| Data(Request/Response   |
- * | . . . . | . . . . | . . . . . . . . . . . . |
- * 0         4         8                         length
- *
- * \param json Json Payload
- * \param[out] len Length of the message
+/*
+ * Prepare byte stream in the proxy protocol format.
  *
- * \return Newly allocated string on sucess, NULL on failure
+ * \param header \ref frame_header
+ * \param payload Payload to be sent
+ * \param len Total size of the header and payload
  */
-char*
-get_proxy_ctl_msg(const char *json, size_t *len) {
-	char   *proxy_ctl_msg = NULL;
-	size_t  json_len;
+uint8_t*
+serialize_message(struct frame_header *header, const uint8_t *payload, ssize_t len)
+{
+	uint8_t *msg = NULL;
 
-	if (! (json && len)) {
+	if (! (header && len)) {
 		return NULL;
 	}
 
-	json_len = strlen(json);
-	*len = json_len + PROXY_CTL_HEADER_SIZE;
-	proxy_ctl_msg = calloc(*len + 1, sizeof(char));
+	msg = calloc((size_t)len, sizeof(uint8_t));
 
-	if (! proxy_ctl_msg) {
+	if (! msg) {
 		abort();
 	}
 
-	set_big_endian_32((uint8_t*)proxy_ctl_msg + PROXY_CTL_HEADER_LENGTH_OFFSET, 
-				(uint32_t)(json_len));
-	strncpy(proxy_ctl_msg + PROXY_CTL_HEADER_SIZE, json, json_len);
+	set_big_endian_16(msg, PROXY_VERSION);
+	/* Header length expressed in 32 bit words.
+	 */
+	msg[HEADER_LEN_OFFSET] = (header->header_len/32);
```

Comment:
Magic number. What happens when the protocol changes to some other value? I'd create a documented `#define` and use that.